src/app.ts

```diff
@@ -0,0 +1,366 @@
+                    hash: contract.deploymentTxHash,
+                })
+                return { ...contract, receipt }
+            }),
+        )
+
+        const proxiesWithRateProviderCompleteInfo = await Promise.all(
+            receipts.map(async (info) => {
+                if (!info.ABI) {
+                    throw new Error(`ABI not provided for contract at address ${info.address}`)
+                }
+                const events = parseEventLogs({
+                    logs: info.receipt.logs,
+                    abi: JSON.parse(info.ABI),
+                    eventName: 'Upgraded',
+                })
+                const wasUpgraded = events.length > 0
+
+                return { ...info, events, wasUpgraded }
+            }),
+        )
+
+        return proxiesWithRateProviderCompleteInfo.map((r) => {
+            if (!r.Implementation) {
+                throw new Error(`Implementation is undefined for contract at address ${r.address}`)
+            }
+            return {
+                address: r.address,
+                wasUpgraded: r.wasUpgraded,
+                implementation: r.Implementation,
+            }
+        })
+    }
+
+    /**
+     * Checks if the rate provider is upgradeable.
+     * @returns True if the rate provider is upgradeable, false otherwise.
+     */
+    public async getIsRateProviderUpgradeable(): Promise<boolean> {
+        const upgradeableContracts = await this.getUpgradeableContracts()
+        return upgradeableContracts.some((c) => c.wasUpgraded)
+    }
+
+    /**
+     * Fetches the source code for the given address.
+     * @param address The address to fetch source code for.
+     * @returns The source code information.
+     */
+    public async getContractSourceCode(
+        address: Address,
+    ): Promise<{ address: Address; Proxy: string; ContractName: string; ABI: string }> {
+        const etherscanApi = new EtherscanApi(this.chain, this.apiKey)
+        const sourceCodeArray = await etherscanApi.getSourceCode([address])
+        return sourceCodeArray[0]
+    }
+
+    /**
+     * Analyzes the getRate call for the rate provider.
+     * @returns An object containing scale18 and rateScale18 information.
+     */
+    public async isRateScale18(): Promise<{ scale18: boolean; rateScale18: bigint }> {
+        const publicClient = createPublicClient({
+            chain: this.chain,
+            transport: http(),
+        })
+
+        const data = (await publicClient.readContract({
+            address: this.rateProvider,
+            abi: rateProviderAbi,
+            functionName: 'getRate',
+            args: [],
+        })) as bigint
+
+        if (this.sourceCode.ContractName === 'ConstantRateProvider') {
+            return {
+                scale18: true,
+                rateScale18: data,
+            }
+        }
+
+        // rate must be scale 18
+        const isScale18 = data < 9n * 10n ** 18n && data >= 1n * 10n ** 18n ? true : false
```

Comment:
I don't follow this check 🤔 
Are you checking if rate is between 1e18 and 9e18?
I've seen rates of 1000e18, so that check would fail.

Reply:
Curious if you recall the types of pools where this is the case. I am aware of the following two scenarios:
- "oracle pool" Where the pools tokens are not alike. Like our USDC-WETH pool on Base with the stableSurge
- Gyros ConstantRateProviders

All other rate providers I recall would fall within this testing scope I think.